Add helper to support include/exclude configuration

aptls opened the pull request:
Add a helper class that can be used by connectors to filter items
before indexing.

Bug: 112165186

Commit:
Add helper to support include/exclude configuration

Add a helper class that can be used by connectors to filter items
before indexing.

Bug: 112165186

donghanmiao commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,164 @@
+ *
+ * <p>Configuration file:
+ * <li>the path to the include/exclude pattern file is defined by the configuration property
+ *     <b>connector.includeExcludeFilter.includeExcludePatternFile</b>
```
why use a different configuration file instead of the existing one?

aptls replied:
The primary reason is that that's how the initial version of this class was implemented. There are some potential advantages to a plain-text file for correctly escaping regular expressions.

donghanmiao replied:
I see. isn't the config file also plain-text file? Having one configuration makes it easier for us to package the connector in a container and pass in parameters as needed.

aptls replied:
The connector config file is read as a Java properties file, so any backslashes in a regular expression have to be doubled there in order to be escaped and passed in to the filter. We'll also have to come up with a way to specify the patterns as named properties. You should be able to list multiple include and exclude patterns, which in the separate file all use the same prefix, while in a properties file we'll need a format like "include.1.regex", include.2.regex", ..., and a way to read these properties.

TanmayVartak replied:
Initial idea to maintain separate file is to allow multiple connector instances use same set of include/exclude patterns.

donghanmiao commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,164 @@
+            excludeRules.add(
+                new Rule<>(new RegexPredicate(line.substring(EXCLUDE_RULE_PREFIX.length()))));
+          } else {
+            logger.log(Level.WARNING, "Invalid Rule [{0}] for include exclude pattern", line);
```
Should this condition error out instead?

donghanmiao commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,164 @@
+    return new IncludeExcludeFilter(includeRules, excludeRules);
+  }
+
+  public boolean isAllowed(String value) {
```
Add javadoc.

Commit:
Add Javadoc, fail on invalid pattern.

Commit:
Use properties from config file instead of separate file for regex patterns.

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,159 @@
+ *   <li>include patterns are prefixed with "includeExcludeFilter.include.regex."
+ *   <li>exclude patterns are prefixed with "includeExcludeFilter.exclude.regex."
+ * </ul>
+ * The last part of the property name is not used here but can be used to describe the
```
"is not used here" seems off. Maybe "is a unique, meaningful name [to describe]"

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,159 @@
+ * The last part of the property name is not used here but can be used to describe the
+ * pattern. Each property name must be unique. For example
+ * <ul>
+ *   <li>includeExcludeFilter.include.regex.textFiles = .*\\.txt
```
This example seems *too* basic, since it won't work for FS and SP. If this worked as expected, the exclude would also be redundant. I think we should document both a real-world "index only these extensions", which would be something like exclude ".*\\.(?!(txt|html))[a-z]*", and a use of the prefix paths.

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,159 @@
+ */
+public class IncludeExcludeFilter {
+  private static final Logger logger = Logger.getLogger(IncludeExcludeFilter.class.getName());
+  public static final String FILTER_CONFIG_PREFIX = "includeExcludeFilter.";
```
If these are public, they need javadoc. Does this one specifically, need to be visible?

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,159 @@
+  public static final String INCLUDE_RULE_PREFIX = FILTER_CONFIG_PREFIX + "include.regex.";
+  public static final String EXCLUDE_RULE_PREFIX = FILTER_CONFIG_PREFIX + "exclude.regex.";
+
+  @VisibleForTesting final List<Rule<String>> includeRules;
```
Can we make these ImmutableLists? It's not that inefficient in the constructor to copy them, either, because ImmutableList.copyOf is cheap when given an ImmutableList.

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,159 @@
+  public boolean isAllowed(String value) {
+    boolean exclude = evaluateRules(excludeRules, value, false /* nothing is excluded */);
+    if (exclude) {
+      logger.log(Level.FINEST, "excluding " + value);
```
Do we want to log not-included stuff as well, or even included stuff (maybe FINER for excluded and not-included, and FINEST for included)?

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,159 @@
+    if (rules.isEmpty()) {
+      return emptyRulesOutcome;
+    }
+    return rules.stream().map(r -> r.eval(value)).filter(e -> e == true).findFirst().orElse(false);
```
Maybe just "e -> e", or "identity()", or "bool -> "bool"?

Isn't "filter(e -> e).findFirst().orElse(false)" the same as "anyMatch(e -> e)"?

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,159 @@
+    return rules.stream().map(r -> r.eval(value)).filter(e -> e == true).findFirst().orElse(false);
+  }
+
+  private static class Rule<T> {
```
This doesn't seem to be carrying much weight. Why not just use Predicate directly?

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,159 @@
+    private final String regex;
+
+    private RegexPredicate(String regex) {
+      this.p = Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
```
The case-insensitivity should be documented.

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,159 @@
+      boolean matches = matcher.matches();
+      logger.log(
+          Level.FINE,
+          "Pattern [{0}] input [{1}] matches Outcome [{2}] with Regex[{3}]",
```
Should we include the descriptive names in the Rule (there's a reason to have it), for better logging here?

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,159 @@
+    @Override
+    public boolean apply(@Nullable String input) {
+      Matcher matcher = p.matcher(input);
+      boolean matches = matcher.matches();
```
Could be combined with the previous line.

What do you think about matches vs find? The former could be unintuitive, and people will often not get any matches, whereas the latter might be prone to false positives. We could config as "...matches.foo = " or "...find.foo = ", rather than "...regex.foo = ", but I'm not too fond of the names in that context. I don't know if people are more likely to grok adding .* as needed, or ^ and $. We need to document this behavior, whatever we decide on.

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,159 @@
+import javax.annotation.Nullable;
+
+/**
+ * Helper utility to check if a particular item should be indexed. Include/exclude
```
"Helper utility" is a very generic start. Maybe "Matches values against configured include/exclude patterns."?

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,159 @@
+ * <ul>
+ *   <li>include patterns are prefixed with "includeExcludeFilter.include.regex."
+ *   <li>exclude patterns are prefixed with "includeExcludeFilter.exclude.regex."
+ * </ul>
```
I'm not sure if this is the right place, but we should add something like "Allowed values much match at least one include pattern, if any are given, and not match any exclude patterns."

Commit:
Code review changes.

Commit:
Use separate rules based on item type. This is intended to provide
easier support for filtering by file extension: if the file extension
rule only applies to content items, the enclosing folders won't be
excluded by the rule. There may be cases where two similar rules are
needed, one for containers and one for content, where one might
have been used in the past.

Add a main method that can be used for testing a set of rules without
running a connector.

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,412 @@
+          throws IllegalArgumentException, IllegalStateException, PatternSyntaxException {
+        // checkState throws IllegalStateException
+        checkState(!Strings.isNullOrEmpty(name), "Rule name is missing");
+        checkState(!Strings.isNullOrEmpty(itemTypeConfig), "Rule item type is missing: " + name);
```
Would "Rule " + name + " item type is missing." read better? Or maybe "Item type is missing for " + name?

Commit:
split regex and prefix rules

Commit:
Add Windows-specific tests

The file prefix filter tests need to be adapted for the platform on
which the tests are being run.

Update Javadoc.

Get main() working after changes to rules.

Commit:
Treat virtual_container as container

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,636 @@
+ * matches any of those rules. Otherwise, if any INCLUDE rules are configured for an item
+ * type, an item must match at least one INCLUDE rule to be included. When no rules are
+ * configured, all items of the given type are allowed. When combining FILE_PREFIX or
+ * URL_PREFIX filters with REGEX filters in INCLUDe rules, an item must match both at
```
s/INCLUDe/INCLUDE/

I think "both" sounds funny, and I think the sentence is clear without it.

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,636 @@
+ *  includeExcludeFilter.includeText.filterPattern = \\.txt$
+ * </pre>
+ *
+ * <p>Include only content within a subfolder.  This rule will include {@code
```
Maybe "within a specified subfolder". I thought it was going to be a generic pattern for any subfolder (i.e., exclude files at the root).

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,636 @@
+ * includeExcludeFilter.pathToInclude.filterPattern = \\\\host\\shareFolder\\folder\\pathToInclude
+ * </pre>
+ *
+ * <p>Include only content within a subfolder of a web site. This rule will include {@code
```
Same here.

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,636 @@
+        .filter(r -> (r.getFilterType().equals(FilterType.FILE_PREFIX)
+                || r.getFilterType().equals(FilterType.URL_PREFIX)))
+        .filter(r -> r.getAction().equals(Action.INCLUDE))
+        .collect(Collectors.toList());
```
Why not construct the ImmutableLists directly with ImmutableList.toImmutableList?

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,636 @@
+    boolean include =
+        evaluateRules(prefixIncludeRules, value, true)
+        &&
+        evaluateRules(regexIncludeRules.get(itemType), value, true);
```
Join with previous line?

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,636 @@
+    return include;
+  }
+
+  private boolean evaluateRules(List<Rule> rules, String value, boolean emptyRulesOutcome) {
```
Since it isn't obvious in this context, please add javadoc about the return value.

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,636 @@
+    if (rules.isEmpty()) {
+      return emptyRulesOutcome;
+    }
+    return rules.stream().map(r -> r.eval(value)).anyMatch(e -> e);
```
The e is not helping me much. Maybe a comment like "Return true if any of the rule evaulations return true."

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,636 @@
+        checkState(!Strings.isNullOrEmpty(actionConfig), "Rule action is missing: " + name);
+
+        // valueOf throws IllegalArgumentException
+        Action action = Action.valueOf(actionConfig.toUpperCase());
```
These calls to toUpperCase on fixed config strings should use Locale.US.

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,636 @@
+    private FilePrefixPredicate(Action action, String prefix) {
+      this.action = action;
+      configuredPrefix = prefix;
+      prefixPath = Paths.get(prefix.toLowerCase());
```
Why are we using toLowerCase here? Does Path.equals not do a case-insensitive comparison when appropriate? Should we be using Files.isSameFile instead?

aptls replied:
If we don't use toLowerCase, this predicate is no longer uniformly case-insensitive like the other predicates; it will depend on the platform where the filter's running. Files.isSameFile is the same as Path.equals when the paths are equal, and if not, according to the Javadoc, may need "to open or access both files", which seemed like more than this class wanted to do.

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,636 @@
+      ImmutableList.Builder<String> pathBuilder = ImmutableList.builder();
+      String path = url.getPath();
+      if (path.isEmpty() || path.equals("/")) {
+        // Just "http://host:port"
```
Maybe s/http/protocol/

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,636 @@
+
+    System.out.println("Enter test value(s)");
+    try (Scanner in = new Scanner(inStream)) {
+      while (true) {
```
Why not while (in.hasNext()) instead of catching NoSuchElementException?

wiarlawd commented on indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java:
```diff
@@ -0,0 +1,636 @@
+    }
+    List<Rule> excludeRules =
+        Stream.concat(prefixExcludeRules.stream(), regexExcludeRules.get(itemType).stream())
+        .collect(Collectors.toList());
```
I think that if you collect this to ImmutableList, then evaluateRules can take an ImmutableList.

wiarlawd reviewed (commented):
Partial review. I haven't looked at the tests yet.

wiarlawd commented on indexing/src/test/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilterTest.java:
```diff
@@ -0,0 +1,1056 @@
+  @Test
+  public void ruleBuilder_regexMissingItemType_throwsException() {
+    thrown.expect(IllegalStateException.class);
+    thrown.expectMessage("Item type");
```
I have a modest urge to move these down to just the build method, but it can be ignored if you don't share it.

wiarlawd commented on indexing/src/test/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilterTest.java:
```diff
@@ -0,0 +1,1056 @@
+        "http://WWW.EXAMPLE.COM/",
+        "http://www.example.com/anyPath");
+    eval(false, rules,
+        "http://www.example.com:80",
```
This one is a little surprising and unfortunate. Is there any builtin facility for getting the default port for a protocol? It may be worth handling at least HTTP and HTTPS specially. I have no idea what the third most common protocol will be, but I doubt it will be very popular at all.

wiarlawd commented on indexing/src/test/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilterTest.java:
```diff
@@ -0,0 +1,1056 @@
+  private void eval(boolean expected, List<IncludeExcludeFilter.Rule> rules, String... values) {
+    for (IncludeExcludeFilter.Rule rule : rules) {
+      for (String s : values) {
+        assertEquals(rule + " = " + s, expected, rule.eval(s));
```
I think it's worth using an ErrorCollector rule here.

wiarlawd commented on indexing/src/test/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilterTest.java:
```diff
@@ -0,0 +1,1056 @@
+        new IncludeExcludeFilter.Rule.Builder("rule6").setFilterType("REGEX")
+        .setAction("EXCLUDE").setFilterPattern("\\.txt.bak$").setItemType("CONTENT_ITEM").build());
+    IncludeExcludeFilter filter = new IncludeExcludeFilter(rules);
+    assertEquals(2, filter.prefixIncludeRules.size());
```
Add the actual lists as messages. Since you're using it a fair bit, maybe an assertSizeEquals helper method.

Commit:
Code review changes.
